contracts/utils/cryptography/SignatureChecker.sol

```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.6.0 <0.9.0;
+
+import "./ECDSA.sol";
+import "../Address.sol";
+import "../../interfaces/IERC1271.sol";
+
+library SignatureChecker {
+    function isValidSignatureNow(address signer, bytes32 hash, bytes memory signature) internal view returns (bool) {
+        if (Address.isContract(signer)) {
+            try IERC1271(signer).isValidSignature(hash, signature) returns (bytes4 magicValue) {
+                return magicValue == IERC1271(signer).isValidSignature.selector;
+            } catch {
+                return false;
+            }
+        } else {
+            return ECDSA.recover(hash, signature) == signer;
```

Comment:
From our recommendation in `ECDSA`:

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5ecd2b850c2acb52b94d8374ac8d22cb543e71c8/contracts/utils/cryptography/ECDSA.sol#L20-L24

We should be providing some other link in the docs, but there is more info in [this article](https://jimmysong.medium.com/faketoshis-nonsense-signature-8700a44536b5).

So, is this function really safe? Isn't it vulnerable to crafting `signature` and `hash` values that will recover to the `signer` but which do not correspond to a real ECDSA signature?

Reply:
> So, is this function really safe? Isn't it vulnerable to crafting signature and hash values that will recover to the signer but which do not correspond to a real ECDSA signature?

Our own ECDSA.recover doesn't enforce that. You can technically use it to recover a hash that was not properly formated. The same goes for a smart wallet 1271 implementation (the code in `isValidSignature`). In both cases, we cannot do anything about it (the hash might be 191, or 712, or anything not yet standardized. The best we can do is adding the warning you mentioned in the interface for 1271.

Reply:
Ok, for some reason I thought this API was more vulnerable to misuse, but now that I'm looking at it again it I can see that it's the same.